src/datagrid/DataGridFilter.tsx

```diff
@@ -0,0 +1,210 @@
+    isOpen: boolean;
+    filterValue: any;
+};
+
+export class DataGridFilter extends React.PureComponent<DataGridFilterProps, DataGridFilterState> {
+    private refParent = React.createRef<HTMLDivElement>();
+    private refChild = React.createRef<HTMLDivElement>();
+
+    static defaultProps = {
+        filterType: FilterType.STR,
+    };
+
+    // Initial state for filter
+    state: DataGridFilterState = {
+        isOpen: false,
+        filterValue: undefined,
+    };
+
+    getFilterValue = () => {
+        return this.state.filterValue;
+    };
+
+    updateFilter = (value: any) => {
+        const {columnName, datagridRef, onFilter} = this.props;
+
+        // get latest data from grid
+        const rows = datagridRef.current!.getAllRows();
+
+        if (onFilter && datagridRef) {
+            this.setState({filterValue: value}, () =>
+                datagridRef.current!.updateRows(onFilter(rows, value, columnName)),
+            );
+        }
+    };
+
+    private handleButtonClick = () => {
+        this.toggle();
+    };
+
+    private toggle() {
+        this.setState(
+            prevState => ({
+                isOpen: !prevState.isOpen,
+            }),
+            this.afterToggle,
+        );
+    }
+
+    private afterToggle = () => {
+        if (this.state.isOpen) {
+            this.subscribeDocumentClick();
+        } else {
+            this.unsubscribeDocumentClick();
+        }
+    };
+
+    private subscribeDocumentClick = () => {
+        window.addEventListener("click", this.handleDocumentClick as any, true);
+    };
+
+    private unsubscribeDocumentClick = () => {
+        window.removeEventListener("click", this.handleDocumentClick as any, true);
+    };
+
+    private handleDocumentClick = (evt: Event) => {
+        if (this.state.isOpen) {
+            const target = (evt.target as any) as HTMLElement;
+            const el = this.refChild.current;
+
+            if (el && typeof el !== "string" && !el.contains(target)) {
+                this.toggle();
+            }
+        }
+    };
+
+    render() {
+        const {isOpen, filterValue} = this.state;
+        const {style, className, filterType, customFilter} = this.props;
+        const FilterBtnClasses = classNames([
+            ClassNames.DATAGRID_FILTER_BUTTON,
+            filterValue && ClassNames.DATAGRID_FILTERED,
+        ]);
+        return (
+            <div ref={this.refParent} className={classNames([ClassNames.CLR_FILTER])} style={{position: "relative"}}>
+                <Button
+                    defaultBtn={false}
+                    key="filterBtn"
+                    className={FilterBtnClasses}
+                    onClick={this.handleButtonClick}
+                    icon={{
+                        shape: filterValue ? "filter-grid-circle" : "filter-grid",
+                        className: ClassNames.ICON_SOLID,
+                    }}
+                />
+                {isOpen && (
+                    <div>
```

Comment:
What do you think if we create a function renderForOpen or something like that to keep this condition more less?

Reply:
done